transport/src/main/java/io/netty/channel/DefaultChannelConfig.java

```diff
@@ -255,16 +275,44 @@ public ChannelConfig setAllocator(ByteBufAllocator allocator) {
         return this;
     }
 
+    @SuppressWarnings("unchecked")
     @Override
-    public RecvByteBufAllocator getRecvByteBufAllocator() {
-        return rcvBufAllocator;
+    public <T extends RecvByteBufAllocator> T getRecvByteBufAllocator() {
+        return (T) rcvBufAllocator;
     }
 
+    /**
+     * {@inheritDoc}
+     * <p>
+     * This method enforces the {@link ChannelMetadata#minMaxMessagesPerRead()}. If these predetermined limits
+     * are not appropriate for your use case consider extending the channel and overriding {@link Channel#metadata()},
+     * or use {@link #setRecvByteBufAllocator(RecvByteBufAllocator, ChannelMetadata)}.
+     */
     @Override
     public ChannelConfig setRecvByteBufAllocator(RecvByteBufAllocator allocator) {
+        return setRecvByteBufAllocator(allocator, channel.metadata());
+    }
+
+    /**
+     * Set the {@link RecvByteBufAllocator} which is used for the channel to allocate receive buffers.
+     * @param allocator the allocator to set.
+     * @param metadata Used to determine the {@link ChannelMetadata#minMaxMessagesPerRead()} if {@code allocator}
+     * is of type {@link MaxMessagesRecvByteBufAllocator}.
+     * @return this
+     */
+    public ChannelConfig setRecvByteBufAllocator(RecvByteBufAllocator allocator, ChannelMetadata metadata) {
         if (allocator == null) {
             throw new NullPointerException("allocator");
         }
+        if (allocator instanceof MaxMessagesRecvByteBufAllocator) {
+            if (metadata == null) {
+                throw new NullPointerException("metadata");
+            }
+            MaxMessagesRecvByteBufAllocator maxMsgAllocator = (MaxMessagesRecvByteBufAllocator) allocator;
```

Comment:
@trustin - This is how I am preserving the 16 max messages for `ServerSockets`/`NioByteStream` and 1 max message otherwise.  WDYT?